### ISSUE TITLE
feat: support riscv64

### ISSUE THREAD

**Issue by ernado** [GoogleContainerTools/distroless#1269]
# riscv64 architecture support for static images

Add support for riscv64 architecture, so riscv64 support can be added for dependant projects that are building on distroless.

> riscv64 should be limited to only the static images and dependencies. This will reduce impact on the build.

The static images support would be enough.

### Current issues
```
..URL: https://snapshot.debian.org/archive/debian/20230314T211539Z/dists/bullseye/main/binary-riscv64/Packages.xz, Status: 404 NOT FOUND
panic: failed (status: 404) to fetch remote file: "https://snapshot.debian.org/archive/debian/20230314T211539Z/dists/bullseye/main/binary-riscv64/Packages.xz"
```

There are no packages snapshot for riscv64, so `updateWorkspaceSnapshots.sh` will fail.

### Reference

- https://wiki.debian.org/RISC-V (in-progress debian riscv64 port)
- Debian-riscv [mailing list message](https://lists.debian.org/debian-riscv/2023/03/msg00001.html) from me
- https://github.com/carlosedp/riscv-bringup/ (attempt to bring riscv64 support to container infra from 2019 with lots of additional info)
- https://github.com/etcd-io/etcd/pull/15490 (etcd PR where distroless image is required)
- https://github.com/GoogleContainerTools/distroless/pull/1268 (my failed attempt to add riscv64 to distroless)

### Upd

I've made a fork with CI/CD for building [distroless riscv64 images](https://github.com/go-riscv/distroless/pkgs/container/distroless%2Fstatic-unstable) on Debian Sid (unstable):

https://github.com/go-riscv/distroless

Looks like I will stick to it until Debian Trixie is released, where riscv64 will be (hopefully) stable, I think it will be in 2025.

**Comment by ernado** [comment]
I've made another attempt to support riscv64, but it requires medium size non-trivial refactoring: https://github.com/GoogleContainerTools/distroless/pull/1270

**Comment by mengzhuo** [comment]
ping
Debian 13 has released
https://github.com/GoogleContainerTools/distroless/issues/1826

**Comment by Mcfloy** [comment]
Unfortunately linked to #1871 and GitHub will think this will resolve it but this won't. As supporting a new architecture will have a bigger impact than just adding it just for Java, another PR will be created after aiming to support: `static`, `base`, `base-nossl`, `cc` and `java25`.

**Comment by loosebazooka** [comment]
I would like to restructure out build before adding any more architectures. It's a bit expensive at the moment as it's a singular gigantic cloud build taking a few hours. Failure is annoying to deal with and we have a lot of tech debt here that needs to be resolved.

Can someone describe to me why riscv64 is so important? Who is using it? Is this for nvidia based SOCs? or something else? Does that mean we want python support too?

**Comment by Mcfloy** [comment]
Yes I was thinking about the change of code that would need to be done before, if you need help feel free to ask 😄 

I think @mengzhuo can provide more information about the usage.

**Comment by mengzhuo** [comment]
> I would like to restructure out build before adding any more architectures. It's a bit expensive at the moment as it's a singular gigantic cloud build taking a few hours. Failure is annoying to deal with and we have a lot of tech debt here that needs to be resolved.
> 
> Can someone describe to me why riscv64 is so important? Who is using it? Is this for nvidia based SOCs? or something else? Does that mean we want python support too?

YES, it's important.

Not just NVIDIA, it's a part of opensource ISA. For cloud aspect, [Alibaba](https://www.xrvm.com/), [Google](https://www.cnx-software.com/2025/10/17/google-open-source-coral-npu-synaptics-sl2610-edge-ai-socs/), [Meta](https://www.reuters.com/business/meta-buy-chip-startup-rivos-ai-effort-source-says-2025-09-30/) and [Canonical](https://canonical.com/blog/alibaba-damo-academy-and-canonical) has already in the game. The [RISE project](https://riseproject.dev/) has more detail on it. RISC-V will had proper server grade platforms (not mobile SoC) next year and k8s (distroless) is crucial for those CNCF applications to run on RISC-V server without pain.

**Comment by lazyparser** [comment]
> Can someone describe to me why riscv64 is so important?

Yes. It's very important. It is the future of computing. Feel free to check the landscape and support matrix for RISC-V.

[1] https://matrix.ruyisdk.org/
[2] https://landscape.riscv.org/

**Comment by luhenry** [comment]
:+1: on the support for riscv64. [RISE](https://riseproject.dev/) can also bring some funding to this effort if it can be of any help. Thanks!

**Comment by loosebazooka** [comment]
I'll add riscv for static/base in the next week. Just finishing up some refactoring

**Comment by polarathene** [comment]
> It's a bit expensive at the moment as it's a singular gigantic cloud build taking a few hours. Failure is annoying to deal with and we have a lot of tech debt here that needs to be resolved.

Out of curiosity where is the bulk of that time spent? Was emulation via QEMU involved? Compilation without a cache? 🤔

Canonical has `chisel` as an alternative (_bit more verbose/DIY currently though_) to Google Distroless, where it can produce the rough equivalent of distroless images rather quickly (_supported packages are already stripped down_).

I built 12 images for the `static`/`base-nossl`/`base`/`cc` variants with `linux/amd64`, `linux/arm64`, and `linux/riscv64` platforms (_Ubuntu as a base rather than Debian though which is probably a deal-breaker?_) in a little over 3 minutes on a 1vCPU 2GB RAM VPS as a build host.

**Comment by loosebazooka** [comment]
dup #1925
